### PR TITLE
use_scikit_prepro

### PR DESCRIPTION
**Pull request by WATLE**
scikitのpreprossessingを使った。
ついでに、classにもx_norm_rangeの項目を追加(regにはあった)

### CODE REVIEW

**Comment by WATLE** [comment]
#140

File: skqulacs/qnn/classifier.py [hunk 1]
```diff
@@ -203,33 +219,20 @@ def cost_func(self, theta, x_scaled, y_scaled):
...
-        for i in range(len(y)):
-            if y.ndim == 1:
-                res[i][y[i]] = 1
-            else:
-                for j in range(len(y[i])):
-                    res[i][y[i][j] + self.scale_y_param[1][j]] = 1
-        return res
+        return [syurui, syurui_rui]
 
     def rev_y_scale(self, y_inr):
```
Comment (ikanago) [inline skqulacs/qnn/classifier.py:234]: このメソッドは `np.argmax` で簡単に書けそうです．
例えば以下のようになります(テストしてないのでうまく動くかは分からないです)．
```python
def inverse_one_hot(y_pred: np.array):
    # Convert probability to one-hot encoding
    class_indices = y_pred.argmax(axis=1)
    y_one_hot = np.identity(y_pred.shape[1])[class_indices]
    return self.enc.inverse_transform(y_one_hot)
```
参考: https://note.nkmk.me/python-numpy-eye-identity-one-hot/

File: skqulacs/qnn/regressor.py [hunk 2]
```diff
@@ -72,16 +73,29 @@ def fit(
...
+                feature_range=(-self.x_norm_range, self.x_norm_range)
+            )
+            x_scaled = self.scale_x_scaler.fit_transform(x_train)
         else:
             x_scaled = x_train
 
         if self.do_y_scale:
-            y_scaled = self._do_y_scale(y_train)
+            self.scale_y_scaler = MinMaxScaler(
+                feature_range=(-self.y_norm_range, self.y_norm_range)
```
[the remark this answers is not in the document]
Comment (ikanago) [reply skqulacs/qnn/regressor.py:96]: `feature_range` を `(-1, 1)` にしているのは `arcsin` に渡すからですね，思い出しました

File: skqulacs/qnn/classifier.py [hunk 3]
```diff
@@ -203,33 +219,20 @@ def cost_func(self, theta, x_scaled, y_scaled):
...
+        # scikitのone-hotを使っても、　種類ごとにmaxの要素を検出してくれるわけではないので、ここはそのままです。
```
[the remark this answers is not in the document]
Comment (WATLE) [reply skqulacs/qnn/classifier.py:235]: 英語でコメントすると自分で見れないのでどうしましょう...

File: skqulacs/qnn/regressor.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (WATLE) [reply skqulacs/qnn/regressor.py:96]: Y_NORM_RANGE=0.7のほうが性能が良いはず

File: skqulacs/qnn/classifier.py (hunk 1 above)
Comment (WATLE) [reply skqulacs/qnn/classifier.py:234]: classifier,一つのQNNで同時に複数個の分類ができます
例えば、[入力データ１] = [犬,オス]  、　[入力データ2] = [猫、メス]  [入力データ3] = [猫、オス]
のように。
これはscikitにあるという理由で実装し、あまり使わないのでろくにテストもされていないが、後方互換性のために消すわけにもいかず、実装を複雑化させています。
この問題に関しては別のIssueを建てます。

Comment (WATLE) [reply skqulacs/qnn/classifier.py:234]: #151

File: skqulacs/qnn/regressor.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (ikanago) [reply skqulacs/qnn/regressor.py:96]: じゃあコンストラクタに "Setting `y_norm_range` to 0.7 improves performance." みたいなことを書いとくとよさそうです．

File: skqulacs/qnn/classifier.py (hunk 1 above)
Comment (WATLE) [reply skqulacs/qnn/classifier.py:234]: 一つのQNNで同時に複数個の分類の分類をやめることにより、np.argmaxを採用します。

**Comment by WATLE** [comment]
logloss-fixの内容もこれに含まれています

**Comment by WATLE** [comment]
ちなみに、one-fotいらなくなりました

File: skqulacs/qnn/regressor.py [hunk 4]
```diff
@@ -163,13 +179,18 @@ def predict(self, x_test: List[List[float]]) -> List[float]:
...
-            x_scaled = _min_max_scaling(x_test, self.scale_x_param)
+            x_scaled = self.scale_x_scaler.transform(x_test)
         else:
             x_scaled = x_test
 
         if self.do_y_scale:
-            y_pred = self._rev_y_scale(self._predict_inner(x_scaled))
+            y_pred = self.scale_y_scaler.inverse_transform(
+                self._predict_inner(x_scaled)
+            )
```
Comment (ikanago) [inline skqulacs/qnn/regressor.py:193]: かなりシンプルになっていいと思います．

File: skqulacs/qnn/classifier.py [hunk 5]
```diff
@@ -156,113 +163,54 @@ def predict(self, x_test: List[List[float]]):
...
-            cost = 0
-            for i in range(len(ysf)):
-                if ysf[i] == 1:
-                    cost -= np.log(ypf[i])
-                else:
-                    cost -= np.log(1 - ypf[i])
-            cost /= len(ysf)
-            return cost
+            y_pred_sm = softmax(y_pred, axis=1)
+            return log_loss(y_scaled, y_pred_sm)
```
Comment (ikanago) [inline skqulacs/qnn/classifier.py:193]: scikit-learn にあるやつで済むならそのほうがよさそうですね．

File: skqulacs/qnn/classifier.py [hunk 6]
```diff
@@ -156,113 +163,54 @@ def predict(self, x_test: List[List[float]]):
...
-        return res
-
     def _cost_func_grad(self, theta, x_scaled, y_scaled):
         self.circuit.update_parameters(theta)
 
-        mto = self._predict_inner(x_scaled).copy()
+        y_pred = self._predict_inner(x_scaled)
+        y_pred_sm = softmax(y_pred, axis=1)
         grad = np.zeros(len(theta))
         for h in range(len(x_scaled)):
```
Comment (ikanago) [inline skqulacs/qnn/classifier.py:205]: `h` よりも `sample_index` とか `current_sample` のように，「何番目のデータを見ている」という意味を込めたほうが読みやすくなると思いました．

File: skqulacs/qnn/classifier.py [hunk 7]
```diff
@@ -156,113 +163,54 @@ def predict(self, x_test: List[List[float]]):
...
-                for k in range(self.scale_y_param[0][j]):
-                    mto[h][hid + k] = np.exp(self.y_exp_ratio * mto[h][hid + k]) / wa
             backobs = Observable(self.n_qubit)
-
-            for i in range(len(y_scaled[0])):
-                if y_scaled[h][i] == 0:
-                    backobs.add_operator(1.0 / (1.0 - mto[h][i]), f"Z {i}")
-                else:
-                    backobs.add_operator(-1.0 / (mto[h][i]), f"Z {i}")
+            for i in range(self.num_class):
```
Comment (ikanago) [inline skqulacs/qnn/classifier.py:207]: ここの `i` も同様に，`dim` とか `current_class` のように「あるデータの何番目の値を見ている」という意味を込めると読みやすいです．

File: skqulacs/qnn/classifier.py [hunk 8]
```diff
@@ -156,113 +163,54 @@ def predict(self, x_test: List[List[float]]):
...
+                aaa = 0.0
```
Comment (ikanago) [inline skqulacs/qnn/classifier.py:208]: `y_pred_sm[h][i] - aaa` が `0.0` に近いとき `aaa` を `1.0` にしているので，`aaa` よりは `expected` とかのほうが適していそうです．

File: tests/test_qnn_classifier.py [hunk 9]
```diff
@@ -29,4 +29,4 @@ def test_classify_iris(solver: str, maxiter: int):
     for kai in range(maxiter):
         qcl.fit(x_train, y_train, 1)
         y_pred = qcl.predict(x_test)
```
Comment (ikanago) [inline tests/test_qnn_classifier.py:31]: この部分なんですが，古いブランチから checkout しているのか修正されていないので
```python
qcl.fit(x_train, y_train, maxiter)
y_pred = qcl.predict(x_test)
```
としておいてください．

File: skqulacs/qnn/classifier.py [hunk 10]
```diff
@@ -156,113 +163,54 @@ def predict(self, x_test: List[List[float]]):
...
-            r = [
-                self.observables[i].get_expectation_value(state)
-                for i in range(self.n_qubit)
-            ]
-            res.append(r)
-        return np.array(res)
+        res = np.zeros((len(x_list), self.num_class))
+        for i in range(len(x_list)):
+            state = self.circuit.run(x_list[i])
+            for j in range(self.num_class):
```
Comment (ikanago) [inline skqulacs/qnn/classifier.py:181]: observable は `self.n_qubit` 個あるのに対して `self.num_class` がこれと異なる場合は範囲外参照になったり冗長になったりします．
なので内側のループは `range(self.n_qubit)` のままでよくて，前の実装のままでよさそうです．

Comment (ikanago) [reply skqulacs/qnn/classifier.py:181]: そもそもなんですが，たとえば `num_class=3`，`n_qubits=4` で学習すると，ある `i` について `y_pred[i]` が `[0.1, 0.0, 0.1, 0.8]` みたいになることがありえて，これは不正解になるので `num_class == n_qubits` とすべきである気がします．この場合 `num_class` を用意する必要はないです．

**Review by ikanago — CHANGES_REQUESTED** [review]
scikit-learn にあるモジュールを使うとかなりすっきりしたと思います．
いくつかコメントしたので確認お願いします 🙏

File: skqulacs/qnn/classifier.py (hunk 10 above)
Comment (WATLE) [reply skqulacs/qnn/classifier.py:181]: num_class == n_qubitsとしたとき、2値分類するときにnum_class =2となりますが、それだと量子2ビットの回路しか使えなくなります。そうすると、回路の複雑性が不足します。

よって、num_class と n_qubitsを別にして、　番号が大きいビットは観測しないことにしています。

num_class <= n_qubits である必要があることは書くべきかもしれません。

Comment (ikanago) [reply skqulacs/qnn/classifier.py:181]: > 回路の複雑性が不足します

なるほど，そういうことだったんですね

このことと `num_class <= n_qubits` の必要性は明記しておいたほうがたしかに親切だと思います．

Comment (ikanago) [reply skqulacs/qnn/classifier.py:181]: もう書いてありましたね，ありがとうございます

**Review by ikanago — APPROVED** [review]
LGTM です!

**Comment by WATLE** [comment]
QCNN関係が手元でテスト通らないんだけどどうしよう

**Comment by ikanago** [comment]
140-use-scikit-prepro の最新のコミットで `make test` すると僕の手元では通りました

**Comment by ikanago** [comment]
フォーマットで CI 落ちてるので make fix したらマージできそうですね

**Comment by ikanago** [comment]
通りましたね
